cedar-policy-validator/src/cedar_schema/err.rs

```diff
@@ -343,86 +360,305 @@ impl Diagnostic for ToJsonSchemaErrors {
+    ReservedSchemaKeyword(ReservedSchemaKeyword),
 }
 
 impl ToJsonSchemaError {
-    pub fn duplicate_keys(key: SmolStr, loc1: Loc, loc2: Loc) -> Self {
-        Self::DuplicateKeys { key, loc1, loc2 }
+    pub(crate) fn duplicate_context(name: impl ToSmolStr, loc1: Loc, loc2: Loc) -> Self {
+        Self::DuplicateContext(DuplicateContext {
+            name: name.to_smolstr(),
+            loc1,
+            loc2,
+        })
+    }
+
+    pub(crate) fn duplicate_decls(decl: impl ToSmolStr, loc1: Loc, loc2: Loc) -> Self {
+        Self::DuplicateDeclarations(DuplicateDeclarations {
+            decl: decl.to_smolstr(),
+            loc1,
+            loc2,
+        })
     }
-    pub fn duplicate_decls(decl: SmolStr, loc1: Loc, loc2: Loc) -> Self {
-        Self::DuplicateDeclarations { decl, loc1, loc2 }
+
+    pub(crate) fn duplicate_namespace(
+        namespace_id: impl ToSmolStr,
+        loc1: Option<Loc>,
+        loc2: Option<Loc>,
+    ) -> Self {
+        Self::DuplicateNamespaces(DuplicateNamespace {
+            namespace_id: namespace_id.to_smolstr(),
+            loc1,
+            loc2,
+        })
     }
-    pub fn duplicate_namespace(namespace_id: SmolStr, loc1: Loc, loc2: Loc) -> Self {
-        Self::DuplicateNameSpaces {
-            namespace_id,
-            loc1: Some(loc1),
-            loc2: Some(loc2),
-        }
+
+    pub(crate) fn duplicate_principal(name: impl ToSmolStr, loc1: Loc, loc2: Loc) -> Self {
+        Self::DuplicatePrincipalOrResource(DuplicatePrincipalOrResource {
+            name: name.to_smolstr(),
+            kind: PR::Principal,
+            loc1,
+            loc2,
+        })
     }
+
+    pub(crate) fn duplicate_resource(name: impl ToSmolStr, loc1: Loc, loc2: Loc) -> Self {
+        Self::DuplicatePrincipalOrResource(DuplicatePrincipalOrResource {
+            name: name.to_smolstr(),
+            kind: PR::Resource,
+            loc1,
+            loc2,
+        })
+    }
+
+    pub(crate) fn no_principal(name: impl ToSmolStr, loc: Loc) -> Self {
+        Self::NoPrincipalOrResource(NoPrincipalOrResource {
+            kind: PR::Principal,
+            name: name.to_smolstr(),
+            loc,
+        })
+    }
+
+    pub(crate) fn no_resource(name: impl ToSmolStr, loc: Loc) -> Self {
+        Self::NoPrincipalOrResource(NoPrincipalOrResource {
+            kind: PR::Resource,
+            name: name.to_smolstr(),
+            loc,
+        })
+    }
+
+    pub(crate) fn reserved_name(name: impl ToSmolStr, loc: Loc) -> Self {
+        Self::ReservedName(ReservedName {
+            name: name.to_smolstr(),
+            loc,
+        })
+    }
+
+    pub(crate) fn reserved_keyword(keyword: impl ToSmolStr, loc: Loc) -> Self {
+        Self::ReservedSchemaKeyword(ReservedSchemaKeyword {
+            keyword: keyword.to_smolstr(),
+            loc,
+        })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Error)]
+#[error("this uses a reserved schema keyword: `{keyword}`")]
```

Comment:
Add a help message here like `"Keywords `Entity`, `Extension`, `Set`, and `Record` cannot be used as common type names"`?